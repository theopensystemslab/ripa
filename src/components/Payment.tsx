import Box from "@material-ui/core/Box";
import Button from "@material-ui/core/Button";
import Grid from "@material-ui/core/Grid";
import TextField from "@material-ui/core/TextField";
import Typography from "@material-ui/core/Typography";
import { useFormik } from "formik";
import React, { useState } from "react";
import FocusWithin from "react-focus-within";

import Messages from "../shared/components/submit-messages";
import Question from "./Question";

interface IPayment {
  fee: number;
}
const Payment: React.FC<IPayment> = ({ fee }) => {
  return (
    <Box bgcolor="#ddd" px={2} py={5} textAlign="center">
      <Typography variant="h5" component="p">
        The planning fee for this project is
      </Typography>
      <Typography variant="h4" component="p">
        <strong>£{fee}</strong>
      </Typography>
    </Box>
  );
};
interface ICardDetails {
  cardholderName?: string;
  month?: number;
  year?: number;
  securityCode?: number;
  cardNumber?: number;
  includeSubmit?: boolean;
}
export const CardDetails: React.FC<ICardDetails> = ({
  includeSubmit = false
}) => {
  const [successMessageVisible, setSuccessMessageVisible] = React.useState(
    false
  );

  const formik = useFormik({
    initialValues: {
      cardNumber: "",
      securityCode: "",
      month: "",
      year: "",
      cardholderName: ""
    },
    onSubmit: (values, { resetForm }) => {
      console.log(JSON.stringify(values, null, 2));
      setSuccessMessageVisible(true);
      setTimeout(() => {
        resetForm();
        setSuccessMessageVisible(false);
      }, 1000);
    }
  });
  return (
    <Box bgcolor="background.paper" py={6}>
      <FocusWithin>
        {({ isFocused, getFocusProps }) => (
          <form
            data-testid="cardDetailsForm"
            onSubmit={formik.handleSubmit}
            {...getFocusProps()}
          >
            <Box maxWidth={440}>
              <Box pb={1}>
                <Question inFocus={isFocused}>Enter card details</Question>
              </Box>
              <Box pb={3}>
                <TextField
                  type="number"
                  value={formik.values.cardNumber}
                  name="cardNumber"
                  fullWidth
                  onChange={formik.handleChange}
                  placeholder="Card number"
                  label="Card number"
                />
                <Box fontSize="body2.fontSize" color="grey.400" pt={1}>
                  Accepted credit and debit card types
                </Box>
              </Box>
              <Box pb={3}>
                <Typography variant="body2" component="div" gutterBottom>
                  Expiry date
                </Typography>
                <Box maxWidth={200}>
                  <Grid
                    container
                    spacing={1}
                    alignItems="flex-end"
                    wrap="nowrap"
                  >
                    <Grid item>
                      <TextField
                        type="number"
                        value={formik.values.month}
                        name="month"
                        onChange={formik.handleChange}
                        placeholder="Month"
                        label="Month"
                      />
                    </Grid>
                    <Grid item>
                      <Box fontSize="24px" fontWeight="400" lineHeight="40px">
                        /
                      </Box>
                    </Grid>
                    <Grid item>
                      <TextField
                        type="number"
                        value={formik.values.year}
                        name="year"
                        onChange={formik.handleChange}
                        placeholder="Year"
                        label="Year"
                      />
                    </Grid>
                  </Grid>
                </Box>
              </Box>
              <Box pb={3}>
                <TextField
                  value={formik.values.cardholderName}
                  name="cardholderName"
                  onChange={formik.handleChange}
                  placeholder="Name"
                  label="Name on card"
                  fullWidth
                />
              </Box>
              <Box pb={5}>
                <TextField
                  value={formik.values.securityCode}
                  name="securityCode"
                  onChange={formik.handleChange}
                  placeholder="security code"
                  label="Card security code"
                />
              </Box>
            </Box>
            {includeSubmit && (
              <Button type="submit" variant="contained" color="primary">
                Save and Continue
              </Button>
            )}
            <div>
              {successMessageVisible ? (
                <Messages
                  type="success"
                  message="Form submitted successfully"
                />
              ) : null}
            </div>
          </form>
        )}
      </FocusWithin>
    </Box>
  );
};

interface IReceipt {
  fee: number;
  paidOn: Date;
  reference: string;
}
export const Receipt: React.FC<IReceipt> = ({ fee, paidOn }) => (
  <Box bgcolor="#ddd" px={3} py={4} textAlign="center">
    <Typography variant="h5">Planning fee paid</Typography>
    <Typography variant="h4" gutterBottom>
      <strong>£{fee}</strong>
    </Typography>
    <Box fontSize="caption.fontSize" textAlign="left">
      It was paid on
      <Typography variant="h6">{paidOn.toLocaleDateString()}</Typography>
    </Box>
  </Box>
);

export default Payment;
