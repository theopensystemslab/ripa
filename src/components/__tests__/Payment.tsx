import { fireEvent, render, waitFor } from "@testing-library/react";
import * as React from "react";

import Payment, { CardDetails, Receipt } from "../Payment";

describe("Payment Component", () => {
  describe("of type payment", () => {
    it("should have text in the dom", () => {
      const { getByText } = render(<Payment fee={204} />);
      expect(
        getByText(/the planning fee for this project is/i)
      ).toBeInTheDocument();
      expect(getByText(/£204/i)).toBeInTheDocument();
    });
    it("should render snapshot", () => {
      const { asFragment } = render(<Payment fee={204} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("of type Card Details", () => {
    it("should render success message when all inputs are entered successfully", async () => {
      const {
        getByPlaceholderText,
        findByPlaceholderText,
        findByText,
        getByTestId
      } = render(<CardDetails />);
      fireEvent.change(getByPlaceholderText(/card number/i), {
        target: { value: "12345678910" }
      });
      fireEvent.change(getByPlaceholderText(/month/i), {
        target: { value: "7" }
      });
      fireEvent.change(getByPlaceholderText(/year/i), {
        target: { value: "2020" }
      });
      fireEvent.change(getByPlaceholderText(/name/i), {
        target: { value: "Test user" }
      });
      fireEvent.change(getByPlaceholderText(/security code/i), {
        target: { value: "225" }
      });
      fireEvent.submit(getByTestId("cardDetailsForm"));
      expect(
        await findByText(/form submitted successfully/i)
      ).toBeInTheDocument();
    });
    it("should render snapshot", () => {
      const { asFragment } = render(<CardDetails />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe("of type Receipt", () => {
    it("should have text in the dom", () => {
      const { getByText } = render(
        <Receipt
          fee={204}
          paidOn={new Date(2020, 0, 1)}
          reference={"0271911328"}
        />
      );
      expect(getByText(/Planning fee paid/i)).toBeInTheDocument();
      expect(getByText(/£204/i)).toBeInTheDocument();
      expect(getByText(/It was paid on/i)).toBeInTheDocument();
      expect(getByText("1/1/2020")).toBeInTheDocument();
    });
    it("should render snapshot", () => {
      const { asFragment } = render(
        <Receipt
          fee={204}
          paidOn={new Date(2020, 0, 1)}
          reference={"0271911328"}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
