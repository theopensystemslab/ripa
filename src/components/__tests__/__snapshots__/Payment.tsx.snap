// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Payment Component of type Card Details should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-308"
  >
    <form
      data-testid="cardDetailsForm"
    >
      <div
        class="MuiBox-root MuiBox-root-309"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom"
        >
          <strong>
            Enter card details
          </strong>
        </h4>
        <div
          class="MuiBox-root MuiBox-root-340"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Card number
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="cardNumber"
                placeholder="Card number"
                type="number"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiBox-root MuiBox-root-398"
          >
            Accepted credit and debit card types
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-399"
        >
          <div
            class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom"
          >
            Expiry date
          </div>
          <div
            class="MuiBox-root MuiBox-root-400"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-wrap-xs-nowrap MuiGrid-align-items-xs-flex-end"
            >
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                    data-shrink="false"
                  >
                    Month
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiInput-input"
                      name="month"
                      placeholder="Month"
                      type="number"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div
                  class="MuiBox-root MuiBox-root-504"
                >
                  /
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                    data-shrink="false"
                  >
                    Year
                  </label>
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiInput-input"
                      name="year"
                      placeholder="Year"
                      type="number"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-505"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Name on card
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="cardholderName"
                placeholder="Name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-506"
        >
          <div
            class="MuiFormControl-root MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Card security code
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="securityCode"
                placeholder="security code"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Save and Continue
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div />
    </form>
  </div>
</DocumentFragment>
`;

exports[`Payment Component of type Receipt should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-583"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5"
    >
      Planning fee paid
    </h5>
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom"
    >
      <strong>
        £204
      </strong>
    </h4>
    <div
      class="MuiBox-root MuiBox-root-614"
    >
      It was paid on
      <h6
        class="MuiTypography-root MuiTypography-h6"
      >
        1/1/2020
      </h6>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Payment Component of type payment should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-34"
  >
    <p
      class="MuiTypography-root MuiTypography-h5"
    >
      The planning fee for this project is
    </p>
    <p
      class="MuiTypography-root MuiTypography-h4"
    >
      <strong>
        £204
      </strong>
    </p>
  </div>
</DocumentFragment>
`;
